internal/datacoord/garbage_collector.go

```diff
@@ -178,103 +179,146 @@ func (gc *garbageCollector) work() {
+// GC the file if checker returns false.
+func (gc *garbageCollector) recycleUnusedBinLogWithChecker(ctx context.Context, prefix string, label string, checker func(objectInfo *storage.ChunkObjectInfo, segment *SegmentInfo) bool) {
+	logger := log.With(zap.String("prefix", prefix))
+	logger.Info("gc start scan", zap.String("prefix", prefix))
+	lastFilePath := ""
+	total := 0
+	valid := 0
+	unexpectedFailure := 0
+	removed := 0
+	start := time.Now()
+
+	err := gc.option.cli.WalkWithPrefix(ctx, prefix, true, func(chunkInfo *storage.ChunkObjectInfo) error {
+		total++
+		lastFilePath = chunkInfo.FilePath
+
+		// Check file tolerance first to avoid unnecessary operation.
+		if time.Since(chunkInfo.ModifyTime) <= gc.option.missingTolerance {
+			logger.Info("skip file since it is not expired", zap.String("filePath", chunkInfo.FilePath), zap.Time("modifyTime", chunkInfo.ModifyTime))
+			return nil
 		}
-		cost := time.Since(startTs)
-		segmentMap, filesMap := getMetaMap()
-		metrics.GarbageCollectorListLatency.
-			WithLabelValues(fmt.Sprint(paramtable.GetNodeID()), labels[idx]).
-			Observe(float64(cost.Milliseconds()))
-		log.Info("gc scan finish list object", zap.String("prefix", prefix), zap.Duration("time spent", cost), zap.Int("keys", len(infoKeys)))
-		for i, infoKey := range infoKeys {
-			total++
-			_, has := filesMap[infoKey]
-			if has {
-				valid++
-				continue
-			}
 
-			segmentID, err := storage.ParseSegmentIDByBinlog(gc.option.cli.RootPath(), infoKey)
-			if err != nil {
-				missing++
-				log.Warn("parse segment id error",
-					zap.String("infoKey", infoKey),
-					zap.Error(err))
-				continue
-			}
+		// Parse segmentID from file path.
+		// TODO: Does all files in the same segment have the same segmentID?
+		segmentID, err := storage.ParseSegmentIDByBinlog(gc.option.cli.RootPath(), chunkInfo.FilePath)
+		if err != nil {
+			unexpectedFailure++
+			logger.Warn("parse segment id error",
+				zap.String("filePath", chunkInfo.FilePath),
+				zap.Error(err))
+			return nil
+		}
 
-			if strings.Contains(prefix, common.SegmentInsertLogPath) &&
-				segmentMap.Contain(segmentID) {
-				valid++
-				continue
-			}
+		segment := gc.meta.GetSegment(segmentID)
+		if checker(chunkInfo, segment) {
+			valid++
+			logger.Info("skip file since it is valid", zap.String("filePath", chunkInfo.FilePath), zap.Int64("segmentID", segmentID))
+			return nil
+		}
 
-			// not found in meta, check last modified time exceeds tolerance duration
-			if time.Since(modTimes[i]) > gc.option.missingTolerance {
-				// ignore error since it could be cleaned up next time
-				removedKeys = append(removedKeys, infoKey)
-				err = gc.option.cli.Remove(ctx, infoKey)
-				if err != nil {
-					missing++
-					log.Error("failed to remove object",
-						zap.String("infoKey", infoKey),
-						zap.Error(err))
-				}
-			}
+		// ignore error since it could be cleaned up next time
+		err = gc.option.cli.Remove(ctx, chunkInfo.FilePath)
```

Comment:
maybe we can have a go routine pool to delete asyncly?

Reply:
It's better option to increase the task throughput.
I would add a new configuration to control the concurrency too.